src/plugins/content_management/server/event_stream/README.md

```diff
@@ -0,0 +1,50 @@
+# Event Stream
+
+
+## The service
+
+On a high-level the Event Stream is exposed through the `EventStreamService`
+class, which is the public interface to the Event Stream, it holds any necessary
+state, and follows plugin life-cycle methods.
+
+The service also validates the events before they are stored. It also buffers
+the events on write. Events are buffered for 250ms or up to 100 events before
+they are flushed to the storage.
+
+
+## The client
+
+On a lower level the actual event storage is defined in the `EventStreamClient`
+interface. There are two `EventStreamClient` implementations:
+
+- `EsEventStreamClient` is the production implementation, which stores events
+  to the Elasticsearch.
+- `MemoryEventStreamClient` is used for testing and could be used for demo
+  purposes.
+
+
+### The `EsEventStreamClient` client
+
+`EsEventStreamClient` is used in production. I stores events in the
+`.kibana-event-stream` data stream. The data stream and index template are
+created during plugin initialization "start" life-cycle.
+
+The mappings define `meta` and `indexed` fields, which are reserved for future
+schema extensions (so that new fields can be added without mapping changes).
+
+The mappings also define a transaction ID (`txID`) field, which can be used to
```

Comment:
Ah, good point!  I had thought of that, before seeing the mappings, and after seeing `txID`, then forgot after I saw how small the mappings are.  And I don't believe we have a good place to put `type` / `id` fields for object/subject, but certainly `txId`.  [`transaction.id`](https://www.elastic.co/guide/en/ecs/8.6/ecs-tracing.html) might fit well there.  

Certainly poke through the event docs, lots of good pre-existing properties available.

Reply:
> And I don't believe we have a good place to put type / id fields for object/subject,

Yeah, we can also [propose](https://github.com/elastic/ecs/pulls) these fields to be included in ECS (now or in the future).

Reply:
Yes, we have thought about that and have mixed feelings.

---

Before the Event Stream, let me briefly mention our experience in the Files Service.

We recently implemented the Files Service and our [Blob Storage Spec](https://docs.google.com/document/d/1BlCQYxHwgGZMxysc0BiWxT9ZCLKwgryhkD9ryDNDIuI/edit#) uses [ECS File field](https://www.elastic.co/guide/en/ecs/current/ecs-file.html) for storing file metadata. So far it worked out well for us, but also we haven't seen any big benefits from using the ECS File field schema either.

Many fields that we use were available in the ECS File field schema, but quite a few are not. So, we are using the convention, that the fields that are not part of the ECS start with the capital letter. We had a task to create an RFC to add some of our fields to the ECS schema, it was not high priority for us, and ECS people were not super enthusiastic about it either; so, we eventually dropped that task.

---

Now regarding the Event Stream: I don't think we necessarily need ECS schema, but we looked if we could use it. From the existing fields there was not really much we could use, we can potentially use `event.action` and `transaction.id`, but that is about it.

The biggest thing here is that Event Stream is all about the *subjects* (who performed the action) and *objects* (on what the action was performed. Both abstract entities, both can be of any type (well usually subjects will be users). And I don't see in the ECS schema any existing fields that map well to this.

I could change:

- `txId` to `transaction.id`;
- `predicate` to `event.action`;
- and `payload` to `event.Payload` (note capital letter "P", because `event` does not have "payload).

But not sure if we want to even set an expectation that we will follow ECS. As the idea of the Event Stream is to be very lean, and to just store the `subject-predicate-object` triples.

It does not even have to be Elasticsearch where we store them. So it is an internal implementation detail of the storage layer, which we should not expose to the outside. And enforcing the ECS schema, sort of, sets the expectation that somebody could then search through that data, basically going right into our implementation detail.

But sure if it is worth it: (1) as it creates an expectation that we might follow the ECS schema further, and maybe add more fields later (which we really don't want to do); (2) it sort of implies that somebody should be able to easily look at the raw Event Stream data (because events will follow ECS schema) and I'm not sure we want to expose the Event Stream as raw data, how it is stored in the storage layer is its implementation detail. If we want the data in a searchable form, there could be some "export" functionality from the Event Stream storage to ECS-like object.

Reply:
I'm not sure `event.*` field fits nicely, maybe. The `transaction.id` LGTM. But to introduce here ECS schema just for `transaction.id`&mdash;I'm not sure about that. Thoughts?

Reply:
> we haven't seen any big benefits from using the ECS File field schema either.

I think we'll see the benefits at a later point with the File Service adoption, when the demand to correlate this data with other ECS compliant sources (e.g. Kibana audit logging that uses ECS) will increase :slightly_smiling_face: 

> it sort of implies that somebody should be able to easily look at the raw Event Stream data (because events will follow ECS schema) and I'm not sure we want to expose the Event Stream as raw data

I admit I don't have enough context, but don't we want our users to eventually build custom dashboards\visualizations\detection rules on top of the event stream data? Or our idea is to always build our own curated UI on top of that?

> and ECS people were not super enthusiastic about it either; so, we eventually dropped that task.

Thanks for this context @vadimkibana, I wasn't aware of that. This sounds extremely concerning to me, to be honest. How do we expect industry to adopt ECS if our own consumers are struggling to adopt and extend it even for a brand new functionality? Either I'm misunderstanding the purpose and use-cases behind ECS (might be the case :slightly_smiling_face:), or something is broken here. 

@thomasneirynck @lukeelmers would it make sense to discuss these issues with ECS team at some point? It would be great to understand if ECS is the right tool for Events Stream / Event Log and other similar functionality we might want to add in the future, and if it's, how we can be more efficient in using and extending it.

> Ya, seems like not worth the effort at this point. Would be easy to extend later if we need to

Yeah, I agree, let's try to clarify our concerns with ECS team first before investing more in it. Anyway, this concern shouldn't block this PR.